vignettes/package.Rmd

```diff
@@ -2,119 +2,250 @@
+}
 
-2.  It allows users to specify a desired location for Python before interacting with your package. For example:
+use_backend <- function(backend, gpu = TRUE) {
+  py_require("tensorflow", action = "remove") # remove default backend
+
+  switch(paste0(backend, "_", get_os()),
+    Linux_jax = if (gpu) py_require("jax[cuda12]") else py_require("jax[cpu]"),
+    macOS_jax = py_require(c("jax", if (gpu) "jax-metal")),
+    Windows_jax = py_require("jax"),
+    Linux_tensorflow = { ... },
+    macOS_tensorflow = { ... },
+    Windows_tensorflow = { ... },
+    Linux_torch = { ... },
+    macOS_torch = { ... },
+    Windows_torch = { ... }
+  )
+}
+```
 
-    ``` r
-    library(mypackage)
-    reticulate::use_virtualenv("~/pythonenvs/userenv")
-    # call functions from mypackage
-    ```
+keras3 users can then specify a backend like this:
+
+``` r
+library(keras3)
+use_backend("jax")
+```
+
+### Best Practices
+
+Be mindful that other R packages and users may also declare Python requirements.
+Avoid restrictive version constraints. If a version constraint is necessary,
+prefer `>=` and `!=` over `<=`, as the latter can quickly become outdated.
+
+An example user script header:
+
+``` r
+library(pysparklyr) # declares requirements for PySpark
+library(keras3)     # declares requirements for default 'tensorflow' backend
+use_backend("jax")  # removes 'tensorflow' requirements, adds 'jax' requirements
+
+library(reticulate)
+py_require(c("scipy", "polars"))         # user-declared requirements
+py_require(python_version = ">=3.12")
+py_require(exclude_newer = "2025-01-19")
+
+np <- import("numpy")  # <-- Python initialized
+...
+```
+
+### Declaring Optional Dependencies
+
+It's recommended that all `py_require()` calls be made before reticulate
+initializes the Python session. However, for rarely used optional dependencies,
+the requirement can be declared right before use:
+
+``` r
+model_to_dot <- function(x, ...) {
+  reticulate::py_require("pydot")
+  keras$utils$model_to_dot(x, ...)
+}
+```
+
+Calling `py_require()` after Python has initialized causes reticulate to
+activate a new ephemeral virtual environment containing the additional
+requirements. Only adding packages is permitted after Python has initialized;
+calling `py_require()` with `action="set"` or `action="remove"` is not possible.
+
+## Delay Loading Python Modules
+
+If your R package wraps Python modules, it's common to import them within
+`.onLoad()`. Use the `delay_load` flag in `import()` to allow:
+
+1.  Successful R package loading even when Python packages are not installed
+    (important for CRAN testing).
+2.  Users to specify their Python installation before using your package.
 
-Without the `delay_load`, Python would be loaded immediately and the user's call to `use_virtualenv` would have no effect.
+Example:
+
+``` r
+scipy <- NULL
+
+.onLoad <- function(libname, pkgname) {
+  reticulate::py_require("scipy")
```

Comment:
Any reason why we can't do something like `scipy <<- py_require("scipy")`, and have that be identical to delay_load below?

Reply:
Unfortunately, there is not a 1:1 correspondence between PyPi Package names and Python module names. Any given PyPi package might install any number of modules, named anything, not necessarily sharing a name with the package. 

Consider the `tensorflow` module, which might be provided by multiple different packages published on PyPi: `tensorflow`, `tensorflow-cpu`, `tensorflow-macos`, `tensorflow-metal`, not to mention version constraints or modifiers that are often included with package names like `tensorflow[and-cuda]>=2.18.*`